<a href="https://colab.research.google.com/github/usman-infinity/AIC_345/blob/main/Brand_Perdiction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Generate a synthetic dataset for car brand prediction with at least 1000 samples, including 'Mileage', 'Age', 'Engine Size', and 'Price' as numerical features, and 'Brand' as the categorical target. Split this data into training and testing sets, then apply label encoding to the 'Brand' target variable. Train a classification model (e.g., XGBClassifier or RandomForestClassifier) on the training data, and evaluate its performance using accuracy, precision, recall, F1-score, and a confusion matrix. Interpret these evaluation metrics and calculate feature importances to understand which features are most influential in predicting car brands. Finally, summarize the overall findings of the classification model, its performance, insights from the confusion matrix, and feature importances.

## Generate Synthetic Dataset for Car Brand Prediction

### Subtask:
Create a synthetic dataset with at least 1000 samples. Include 'Mileage', 'Age', 'Engine Size', and 'Price' as numerical features, and 'Brand' (e.g., 'Toyota', 'Honda', 'BMW', 'Mercedes') as the categorical target variable. Ensure plausible relationships between these features and the car brand, with some random noise.


**Reasoning**:
To begin generating the synthetic dataset, I need to import the necessary libraries, define the car brands, and set the number of samples as instructed. This sets up the foundational elements for the data generation process.



In [13]:
import numpy as np
import pandas as pd

# Define a list of car brands
brands = ['Toyota', 'Honda', 'BMW', 'Mercedes']

# Set the number of samples for the dataset
n_samples = 1000

print(f"Libraries imported: numpy, pandas")
print(f"Car brands defined: {brands}")
print(f"Number of samples set: {n_samples}")

Libraries imported: numpy, pandas
Car brands defined: ['Toyota', 'Honda', 'BMW', 'Mercedes']
Number of samples set: 1000


**Reasoning**:
Now that the foundational elements (`brands`, `n_samples`) are defined, I will proceed to generate the synthetic car data. This involves iterating `n_samples` times, randomly selecting a brand, and then generating feature values (Mileage, Age, Engine Size, Price) based on brand-specific plausible distributions with added noise, as specified in the instructions. Finally, I will convert the collected data into a Pandas DataFrame.



In [14]:
car_data = []

# Define parameters for each brand to ensure plausible relationships and noise
brand_params = {
    'Toyota': {'Mileage': (70000, 15000), 'Age': (5, 2), 'Engine Size': (1.8, 0.3), 'Price': (20000, 5000)},
    'Honda': {'Mileage': (65000, 12000), 'Age': (4, 1.5), 'Engine Size': (1.7, 0.2), 'Price': (22000, 4500)},
    'BMW': {'Mileage': (50000, 10000), 'Age': (3, 1), 'Engine Size': (2.5, 0.5), 'Price': (45000, 10000)},
    'Mercedes': {'Mileage': (48000, 9000), 'Age': (3, 1), 'Engine Size': (2.8, 0.6), 'Price': (50000, 12000)}
}

for _ in range(n_samples):
    brand = np.random.choice(brands)
    params = brand_params[brand]

    # Generate features with randomness, ensuring plausible ranges
    mileage = max(1000, np.random.normal(params['Mileage'][0], params['Mileage'][1]))
    age = max(0.5, np.random.normal(params['Age'][0], params['Age'][1]))
    engine_size = max(1.0, np.random.normal(params['Engine Size'][0], params['Engine Size'][1]))
    price = max(5000, np.random.normal(params['Price'][0], params['Price'][1]))

    car_data.append({
        'Brand': brand,
        'Mileage': round(mileage, 0),
        'Age': round(age, 1),
        'Engine Size': round(engine_size, 2),
        'Price': round(price, 2)
    })

# Convert the list of dictionaries into a Pandas DataFrame
df_cars = pd.DataFrame(car_data)

print(f"Synthetic dataset created with {len(df_cars)} samples.")
print("First 5 rows of the dataset:")
print(df_cars.head())
print("\nDataset info:")
df_cars.info()

Synthetic dataset created with 1000 samples.
First 5 rows of the dataset:
      Brand  Mileage  Age  Engine Size     Price
0     Honda  65964.0  3.0         1.84  22618.84
1     Honda  46936.0  4.6         1.69  28545.95
2  Mercedes  51444.0  3.0         2.60  52818.98
3  Mercedes  46649.0  1.7         3.09  26607.54
4  Mercedes  45864.0  2.5         2.08  64463.92

Dataset info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Brand        1000 non-null   object 
 1   Mileage      1000 non-null   float64
 2   Age          1000 non-null   float64
 3   Engine Size  1000 non-null   float64
 4   Price        1000 non-null   float64
dtypes: float64(4), object(1)
memory usage: 39.2+ KB


## Implement and Train a Classification Model

### Subtask:
Train an appropriate classification model (e.g., XGBClassifier or RandomForestClassifier) using `scikit-learn` on your training data to predict the 'Brand'.


**Reasoning**:
I need to import the `RandomForestClassifier` from `sklearn.ensemble`, instantiate it, and then train the model using the `X_train` and `y_train_encoded` datasets as per the instructions.



In [15]:
from sklearn.ensemble import RandomForestClassifier

# Instantiate the RandomForestClassifier model
# Using a random_state for reproducibility
model = RandomForestClassifier(random_state=42)

# Train the model
model.fit(X_train, y_train_encoded)

print("RandomForestClassifier model trained successfully.")

RandomForestClassifier model trained successfully.


## Evaluate Classification Model Performance

### Subtask:
Make predictions on your test set. Calculate and print relevant classification metrics such as accuracy, precision, recall, F1-score, and a confusion matrix to evaluate the model's performance in predicting car brands.


**Reasoning**:
To evaluate the model's performance, I need to make predictions on the test set (`X_test`) using the trained `model` and then calculate various classification metrics and the confusion matrix. I will use `sklearn.metrics` for this purpose.



In [16]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix

# Make predictions on the test set
y_pred_encoded = model.predict(X_test)

# Calculate evaluation metrics
accuracy = accuracy_score(y_test_encoded, y_pred_encoded)
precision = precision_score(y_test_encoded, y_pred_encoded, average='weighted', zero_division=0)
recall = recall_score(y_test_encoded, y_pred_encoded, average='weighted', zero_division=0)
f1 = f1_score(y_test_encoded, y_pred_encoded, average='weighted', zero_division=0)

# Calculate the confusion matrix
conf_matrix = confusion_matrix(y_test_encoded, y_pred_encoded)

print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-Score: {f1:.4f}")
print("\nConfusion Matrix:\n", conf_matrix)

Accuracy: 0.7600
Precision: 0.7596
Recall: 0.7600
F1-Score: 0.7598

Confusion Matrix:
 [[31  0  6  0]
 [ 1 45  0 17]
 [ 6  0 29  0]
 [ 0 18  0 47]]


## Interpret Evaluation Metrics and Calculate Feature Importances

### Subtask:
Interpret the calculated evaluation metrics (accuracy, precision, recall, F1-score) and the confusion matrix. Also, calculate and display the feature importances from the trained classification model to understand which features ('Mileage', 'Age', 'Engine Size', 'Price') are most influential in predicting car brands.

**Reasoning**:
To interpret the model's performance, I will first analyze the computed evaluation metrics (accuracy, precision, recall, and F1-score) and the confusion matrix to understand how well the model performed across different brands. Then, I will calculate and display the feature importances from the `RandomForestClassifier` model to identify which input features (`Mileage`, `Age`, `Engine Size`, `Price`) contributed most significantly to the predictions.



In [17]:
print("\n--- Interpretation of Evaluation Metrics ---")
print(f"Accuracy: {accuracy:.4f} - This indicates the overall proportion of correctly classified instances. An accuracy of {accuracy:.4f} means that {accuracy*100:.2f}% of the car brands were predicted correctly.")
print(f"Precision: {precision:.4f} - This measures the proportion of positive identifications that were actually correct. A precision of {precision:.4f} suggests that when the model predicts a certain brand, it is correct {precision*100:.2f}% of the time on average, weighted by support.")
print(f"Recall: {recall:.4f} - This measures the proportion of actual positives that were identified correctly. A recall of {recall:.4f} means that the model identified {recall*100:.2f}% of all actual car brands correctly on average, weighted by support.")
print(f"F1-Score: {f1:.4f} - The F1-Score is the harmonic mean of precision and recall, providing a balanced measure. An F1-Score of {f1:.4f} suggests a reasonably good balance between precision and recall for the predictions.")


print("\n--- Interpretation of Confusion Matrix ---")
print("The confusion matrix shows the number of correct and incorrect predictions made by the classification model compared to the actual outcomes. Rows represent actual classes, and columns represent predicted classes.")
# Map encoded labels back to original brands for better interpretation if label_encoder was available
# Since it's not directly available in current kernel state, we'll interpret based on indices assuming order
# Brands: ['Toyota', 'Honda', 'BMW', 'Mercedes'] -> Encoded: [0, 1, 2, 3] (assuming alphabetical or input order)
# Given the confusion matrix structure and variable brands, let's assume mapping:
# 0: BMW (from y_train_encoded array, 0 often corresponds to the first unique value)
# 1: Honda
# 2: Mercedes
# 3: Toyota

print("Assuming encoded brands mapping: 0=BMW, 1=Honda, 2=Mercedes, 3=Toyota (based on common label encoding order)")
print("Confusion Matrix:")
print(conf_matrix)
print("\n- Diagonal elements (e.g., [0,0]=31, [1,1]=45) show the number of correct predictions for each brand.")
print("- Off-diagonal elements show misclassifications. For example, [1,3]=17 means 17 actual Hondas were predicted as Toyotas.")
print("- The matrix indicates that while some brands like BMW (0), Honda (1) and Mercedes (2) are reasonably well classified, there are notable misclassifications, e.g., Hondas (1) are sometimes confused with Mercedes (2) and Toyotas (3). Also, Mercedes (2) is sometimes confused with BMW (0). It looks like the model struggles differentiating some brands, leading to off-diagonal entries.")

# Calculate feature importances
feature_importances = model.feature_importances_

# Get feature names from X_train (assuming X_train columns are the features)
feature_names = X_train.columns

# Create a DataFrame for better visualization
feature_importance_df = pd.DataFrame({
    'Feature': feature_names,
    'Importance': feature_importances
}).sort_values(by='Importance', ascending=False)

print("\n--- Feature Importances ---")
print("Feature importances indicate the relative contribution of each feature to the model's prediction accuracy.")
print(feature_importance_df)
print("\n- From the feature importance, we can see which features the model considered most influential in predicting the car brand. Higher importance means the feature played a more significant role.")


--- Interpretation of Evaluation Metrics ---
Accuracy: 0.7600 - This indicates the overall proportion of correctly classified instances. An accuracy of 0.7600 means that 76.00% of the car brands were predicted correctly.
Precision: 0.7596 - This measures the proportion of positive identifications that were actually correct. A precision of 0.7596 suggests that when the model predicts a certain brand, it is correct 75.96% of the time on average, weighted by support.
Recall: 0.7600 - This measures the proportion of actual positives that were identified correctly. A recall of 0.7600 means that the model identified 76.00% of all actual car brands correctly on average, weighted by support.
F1-Score: 0.7598 - The F1-Score is the harmonic mean of precision and recall, providing a balanced measure. An F1-Score of 0.7598 suggests a reasonably good balance between precision and recall for the predictions.

--- Interpretation of Confusion Matrix ---
The confusion matrix shows the number of correc

## Summarize Overall Findings

### Subtask:
Summarize the overall findings of the classification model, its performance, insights from the confusion matrix, and feature importances.


**Reasoning**:
The subtask requires summarizing the overall findings from the model's performance, confusion matrix, and feature importances. I will compile the insights derived from the previously executed steps into a clear, concise summary.



In [18]:
print("\n--- Overall Model Summary ---")
print("The RandomForestClassifier model achieved a satisfactory performance in predicting car brands on the synthetic dataset.")

print("\n**Performance Metrics:**")
print(f"- **Accuracy:** {accuracy:.4f} (meaning {accuracy*100:.2f}% of car brands were correctly predicted overall).")
print(f"- **Precision:** {precision:.4f} (indicating that {precision*100:.2f}% of the model's positive predictions were correct on average).")
print(f"- **Recall:** {recall:.4f} (showing that the model identified {recall*100:.2f}% of all actual car brands correctly on average).")
print(f"- **F1-Score:** {f1:.4f} (representing a good balance between precision and recall).")

print("\n**Insights from Confusion Matrix:**")
print("The confusion matrix revealed that while the model generally performs well, there are specific areas of misclassification. For instance, brands like Honda and Mercedes sometimes get confused with other brands (e.g., Honda with Toyota, and Mercedes with BMW), suggesting that these brands might have overlapping feature characteristics that make them harder to distinguish for the model.")

print("\n**Feature Importances:**")
print("The analysis of feature importances highlighted the most influential factors in predicting car brands:")
print(feature_importance_df.to_string(index=False))
print("\n- **Price** is by far the most significant feature, indicating that it plays a dominant role in differentiating car brands.")
print("- **Engine Size** is the second most important feature, followed by **Mileage** and **Age**.")
print("\nThis suggests that car brand prediction in this synthetic dataset is largely driven by the car's price and engine specifications, with mileage and age having lesser, but still notable, impact.")



--- Overall Model Summary ---
The RandomForestClassifier model achieved a satisfactory performance in predicting car brands on the synthetic dataset.

**Performance Metrics:**
- **Accuracy:** 0.7600 (meaning 76.00% of car brands were correctly predicted overall).
- **Precision:** 0.7596 (indicating that 75.96% of the model's positive predictions were correct on average).
- **Recall:** 0.7600 (showing that the model identified 76.00% of all actual car brands correctly on average).
- **F1-Score:** 0.7598 (representing a good balance between precision and recall).

**Insights from Confusion Matrix:**
The confusion matrix revealed that while the model generally performs well, there are specific areas of misclassification. For instance, brands like Honda and Mercedes sometimes get confused with other brands (e.g., Honda with Toyota, and Mercedes with BMW), suggesting that these brands might have overlapping feature characteristics that make them harder to distinguish for the model.

**Feat

## Summary:

### Data Analysis Key Findings
*   A synthetic dataset was successfully generated with 1000 samples, including numerical features 'Mileage', 'Age', 'Engine Size', 'Price', and a categorical target 'Brand' (Toyota, Honda, BMW, Mercedes), ensuring plausible relationships and randomness.
*   A RandomForestClassifier model was trained to predict car brands.
*   The model achieved a good overall performance with an Accuracy of 0.7600, Precision of 0.7596, Recall of 0.7600, and an F1-Score of 0.7598 on the test set.
*   The confusion matrix revealed specific misclassifications; for example, 17 actual Hondas were incorrectly predicted as Toyotas, and 6 actual Mercedes were misclassified as BMWs. This suggests that certain brands have overlapping characteristics that make them harder to distinguish.
*   Feature importance analysis showed that 'Price' was the most significant feature in predicting car brands (importance: 0.508), followed by 'Engine Size' (importance: 0.266). 'Mileage' (importance: 0.130) and 'Age' (importance: 0.096) had lesser, but still notable, impacts.

### Insights or Next Steps
*   The synthetic dataset highlights that car 'Price' and 'Engine Size' are primary drivers for distinguishing car brands, suggesting these features should be carefully considered in real-world car brand classification tasks.
*   Further investigation into the misclassified samples (e.g., Hondas predicted as Toyotas, Mercedes as BMWs) could involve generating more distinctive features for these brands in the synthetic data or exploring more complex model architectures if using real data.
